src/wallet/wallet_tx_builder.cpp

```diff
@@ -0,0 +1,682 @@
+// Copyright (c) 2022 The Zcash developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or https://www.opensource.org/licenses/mit-license.php .
+
+#include "wallet/wallet_tx_builder.h"
+
+using namespace libzcash;
+
+PrepareTransactionResult WalletTxBuilder::PrepareTransaction(
+        const ZTXOSelector& selector,
+        SpendableInputs& spendable,
+        const std::vector<Payment>& payments,
+        TransactionStrategy strategy,
+        CAmount fee,
+        uint32_t anchorConfirmations) const
+{
+    assert(!payments.empty());
+    assert(selector.RequireSpendingKeys());
+    assert(fee >= 0);
+
+    auto selected = ResolveInputsAndPayments(spendable, payments, strategy, fee);
+    if (std::holds_alternative<InputSelectionError>(selected)) {
+        return std::get<InputSelectionError>(selected);
+    }
+
+    auto resolvedPayments = std::get<std::pair<SpendableInputs, Payments>>(selected).second;
+
+    auto sendFromAccount = wallet.FindAccountForSelector(selector).value_or(ZCASH_LEGACY_ACCOUNT);
+    auto allowedChangeTypes = [&](const std::set<ReceiverType>& receiverTypes) -> std::set<OutputPool> {
+        std::set<OutputPool> result{resolvedPayments.GetRecipientPools()};
+        if (sendFromAccount != ZCASH_LEGACY_ACCOUNT) {
+            result.insert(OutputPool::Sapling);
+        }
+        for (ReceiverType rtype : receiverTypes) {
+            switch (rtype) {
+                case ReceiverType::P2PKH:
+                case ReceiverType::P2SH:
+                    // TODO: This is the correct policy, but it’s a breaking change from previous
+                    //       behavior, so enable it separately.
+                    // if ((spendable.utxos.empty() && strategy.AllowRevealedRecipients())
```

Comment:
This seems wrong (inverting the sense of `spendable.utxos.empty()`).

Reply:
Ah, I think actually the _uncommented_ one is wrong.

The logic in the other pools is like “only if we’re spending from the pool or we’ve AllowedRevealedAmounts.” The commented-out logic for taddrs is like “if we’re not spending utxos, we need AllowRevealedRecipients; if we are spending utxos, we need AllowFullyTransparent.” I think that logic is correct.

The abbreviated uncommented version currently says “only allow transparent change if we’re spending from a taddr (but don’t check AllowFullyTransparent).” I think it should change to either
1. preserve the first part of the original clause (`spendable.utxos.empty() && strategy.AllowRevealedRecipients()`); bu if that causes tests to fail, then
2. always allow transparent change, unconditionally.

There are two follow-up PRs that affect this (not that we shouldn’t get it right here). The first uses the uncommented versions which is stricter about requiring `AllowFullyTransparent` when there is a transparent sender with transparent change. The second simplifies the logic, by using `PrivacyPolicyMeet` to report the correct policy and never explicitly checking `AllowFullyTransparent` (but ensuring that we check `AllowRevealedSenders` on the spendable side).